implementations/rust/ockam/ockam_identity/tests/message_flow_auth.rs

```diff
@@ -0,0 +1,316 @@
+    fn get_channel(&self) -> Address {
+        self.identity
+            .secure_channel_registry()
+            .get_channel_list()
+            .first()
+            .unwrap()
+            .encryptor_messaging_address()
+            .clone()
+    }
+}
+
+async fn create_secure_channel_listener(
+    ctx: &Context,
+    session: &Option<(Sessions, SessionId)>,
+) -> Result<SecureChannelListenerInfo> {
+    let identity = Identity::create(ctx, &Vault::create()).await?;
+
+    let trust_options =
+        SecureChannelListenerTrustOptions::new().with_trust_policy(TrustEveryonePolicy);
+    let trust_options = if let Some((sessions, session_id)) = session {
+        trust_options.with_session(sessions, session_id)
+    } else {
+        trust_options
+    };
+
+    identity
+        .create_secure_channel_listener("listener", trust_options)
+        .await?;
+
+    let info = SecureChannelListenerInfo { identity };
+
+    Ok(info)
+}
+
+struct SecureChannelInfo {
+    identity: Identity<Vault, InMemoryStorage>,
+    address: Address,
+}
+
+async fn create_secure_channel(
+    ctx: &Context,
+    connection: &TcpConnectionInfo,
+) -> Result<SecureChannelInfo> {
+    let identity = Identity::create(ctx, &Vault::create()).await?;
+
+    let trust_options = SecureChannelTrustOptions::new();
+    let trust_options = if let Some((sessions, session_id)) = &connection.session {
+        trust_options.with_ciphertext_session(sessions, session_id)
+    } else {
+        trust_options
+    };
+
+    let address = identity
+        .create_secure_channel(
+            route![connection.address.clone(), "listener"],
+            trust_options,
+        )
+        .await?;
+
+    let info = SecureChannelInfo { identity, address };
+
+    Ok(info)
+}
+
+#[allow(non_snake_case)]
+#[ockam_macros::test]
+async fn sessions__secure_channel_over_tcp_without_session__should_pass_messages(
+    ctx: &mut Context,
+) -> Result<()> {
+    let bob_tcp_info = create_tcp_listener(ctx, false).await?;
+
+    let bob_listener_info = create_secure_channel_listener(ctx, &bob_tcp_info.session).await?;
+
+    let connection_to_bob = create_connection(ctx, &bob_tcp_info.socket_addr, false).await?;
+    ctx.sleep(Duration::from_millis(50)).await; // Wait for workers to add themselves to the registry
+    let connection_to_alice = bob_tcp_info.get_connection();
+
+    check_message_flow(ctx, route![connection_to_bob.address.clone()], true).await?;
+    check_message_flow(ctx, route![connection_to_alice.clone()], true).await?;
+
+    let channel_to_bob = create_secure_channel(ctx, &connection_to_bob).await?;
+    ctx.sleep(Duration::from_millis(50)).await; // Wait for workers to add themselves to the registry
+    let channel_to_alice = bob_listener_info.get_channel();
+
+    check_message_flow(ctx, route![channel_to_bob.address.clone()], true).await?;
+    check_message_flow(ctx, route![channel_to_alice.clone()], true).await?;
+
+    ctx.stop().await
```

Comment:
Aside: I don't understand why the `ctx` is not always stopped by the test macro, and not only when there is a panic or an error. In 99% of the cases that's what we want, isn't it?

Reply:
I remember there was a reason, but I don't remember what it was

Reply:
Maybe you could try it, if you want

Reply:
Ok, I will, at some stage